snuba/migrations/snuba_migrations/transactions/0002_transactions_onpremise_fix_orderby_and_partitionby.py

```diff
@@ -0,0 +1,117 @@
+from snuba.clusters.storage_sets import StorageSetKey
+from snuba.migrations import migration, operations
+
+
+def recreate_table() -> None:
+    """
+    The sample by clause for the transactions table was added in April 2020. Partition
+    by has been changed twice. If the user has a table without the sample by clause,
+    or the old partition by we need to recreate the table correctly and copy the
+    data over before deleting the old table.
+    """
+    cluster = get_cluster(StorageSetKey.TRANSACTIONS)
+
+    clickhouse = cluster.get_query_connection(ClickhouseClientSettings.MIGRATE)
+    database = cluster.get_database()
+
+    table_name = "transactions_local"
+    table_name_new = "transactions_local_new"
+
+    new_sampling_key = "cityHash64(span_id)"
+    new_partition_key = "(retention_days, toMonday(finish_ts))"
+
+    ((curr_sampling_key, curr_partition_key),) = clickhouse.execute(
+        f"SELECT sampling_key, partition_key FROM system.tables WHERE name = '{table_name}' AND database = '{database}'"
+    )
+
+    sampling_key_needs_update = curr_sampling_key != new_sampling_key
+    partition_key_needs_update = curr_partition_key != new_partition_key
+
+    if not sampling_key_needs_update and not partition_key_needs_update:
+        # Already up to date
+        return
+
+    # Create transactions_local_new and insert data
+    ((curr_create_table_statement,),) = clickhouse.execute(
+        f"SHOW CREATE TABLE {database}.{table_name}"
+    )
+
+    new_create_table_statement = curr_create_table_statement.replace(
+        table_name, table_name_new
+    )
+
+    # Insert sample clause before TTL
+    if sampling_key_needs_update:
+        assert "SAMPLE BY" not in new_create_table_statement
+        idx = new_create_table_statement.find("TTL")
+
+        new_create_table_statement = (
+            new_create_table_statement[:idx]
+            + f"SAMPLE BY {new_sampling_key} "
+            + new_create_table_statement[idx:]
+        )
+
+    # Switch the partition key
+    if partition_key_needs_update:
+        assert new_create_table_statement.count(curr_partition_key) == 1
+        new_create_table_statement = new_create_table_statement.replace(
+            curr_partition_key, new_partition_key
+        )
+
+    clickhouse.execute(new_create_table_statement)
+
+    # Copy over data in batches of 100,000
+    [(row_count,)] = clickhouse.execute(f"SELECT count(*) FROM {table_name}")
+    batch_size = 100000
+    batch_count = math.ceil(row_count / batch_size)
+
+    orderby = "toStartOfDay(finish_ts), project_id, event_id"
+
+    for i in range(batch_count):
+        skip = batch_size * i
+        clickhouse.execute(
+            f"""
+            INSERT INTO {table_name_new}
+            SELECT * FROM {table_name}
+            ORDER BY {orderby}
+            LIMIT {batch_size}
+            OFFSET {skip};
+            """
+        )
+
+    clickhouse.execute(f"DROP TABLE {table_name};")
```

Comment:
I would preserve the old table in case of issues and let the user delete it. Also that would allow you to do a reverse of this migration.

Reply:
I dislike the idea of requiring the user to manually delete the table - isn't that what this framework is for? Also, this migration is only for on premise users that are relying on migrations to be automated. They likely haven't manually done any migrations before. There is a risk in dropping the table, but I think it's pretty unlikely if you get all the way to this point successfully.

Reply:
Ok, then I would at least do a sanity check before dropping:
1) select and ensure that the size of the two tables is the same
2) drop the old table as last step after you have renamed the new one. So nothing can fail anymore and a backwards migration would be harmless.

Reply:
Sounds good